diff --git a/libpbe/src/SmtpClient.cc b/libpbe/src/SmtpClient.cc
index 0a4933d..b16ebf9 100644
--- a/libpbe/src/SmtpClient.cc
+++ b/libpbe/src/SmtpClient.cc
@@ -15,6 +15,7 @@
 // You should have received a copy of the GNU General Public License
 // along with this program; if not, write to the Free Software
 // Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+#include <stdio.h>
 #include "SmtpClient.hh"
 
 #include "ip.hh"


diff --git a/libpbe/src/URI.cc b/libpbe/src/URI.cc
index 4019342..834e951 100644
--- a/libpbe/src/URI.cc
+++ b/libpbe/src/URI.cc
@@ -19,10 +19,7 @@
 #include "URI.hh"
 
 #include <boost/spirit.hpp>
-#include <boost/spirit/actor/insert_at_actor.hpp>
-#include <boost/spirit/dynamic/if.hpp>
-#include <boost/spirit/utility/loops.hpp>
-#include <boost/spirit/iterator/multi_pass.hpp>
+#include <boost/spirit/include/classic.hpp>
 
 #include <boost/lexical_cast.hpp>
 #include <boost/scoped_array.hpp>
diff --git a/libpbe/src/parse_http_request.cc b/libpbe/src/parse_http_request.cc
index 8732323..91cb0a7 100644
--- a/libpbe/src/parse_http_request.cc
+++ b/libpbe/src/parse_http_request.cc
@@ -19,10 +19,9 @@
 #include "parse_http_request.hh"
 
 #include <boost/spirit.hpp>
-#include <boost/spirit/actor/insert_at_actor.hpp>
-#include <boost/spirit/dynamic/if.hpp>
-#include <boost/spirit/utility/loops.hpp>
-#include <boost/spirit/iterator/multi_pass.hpp>
+#include <boost/spirit/include/classic.hpp>
+#include <boost/spirit/include/classic.hpp>
+#include <boost/spirit/home/classic/actor.hpp>
 
 #include <boost/lexical_cast.hpp>
 #include <boost/scoped_array.hpp>
