diff --git a/src/rtapi/uspace_rtapi_app.cc b/src/rtapi/uspace_rtapi_app.cc
index 69a71e6..b6d9863 100644
--- a/src/rtapi/uspace_rtapi_app.cc
+++ b/src/rtapi/uspace_rtapi_app.cc
@@ -504,10 +504,10 @@ int main(int argc, char **argv) {
         if(fallback_uid == 0)
         {
             fprintf(stderr,
-                "Refusing to run as root without fallback UID specified\n"
+                "WARNING run as root without fallback UID specified !!\n"
                 "To run under a debugger with I/O, use e.g.,\n"
                 "    sudo env RTAPI_UID=`id -u` RTAPI_FIFO_PATH=$HOME/.rtapi_fifo gdb " EMC2_BIN_DIR "/rtapi_app\n");
-            exit(1);
+            // exit(1);
         }
         setreuid(fallback_uid, 0);
         fprintf(stderr,
