diff --git a/src/libexpr/primops.cc b/src/libexpr/primops.cc
index 0d4efc4..8beb3f7 100644
--- a/src/libexpr/primops.cc
+++ b/src/libexpr/primops.cc
@@ -1021,6 +1021,30 @@ static void prim_unsafeDiscardOutputDependency(EvalState & state, Value * * args
 }
 
 
+static void prim_hash(EvalState & state, Value * * args, Value & v)
+{
+    PathSet context;
+
+    string s = state.coerceToString(*args[0], context);
+    string type = state.coerceToString(*args[1], context);
+
+    HashType ht;
+    if (type == "md5"){
+      ht = htMD5;
+    } else if (type == "sha256"){
+      ht = htSHA256;
+    } else {
+      throw Error(format("bad hash type `%1%'") % type);
+    }
+
+    Hash h = hashString(htMD5, s);
+
+    string hash = printHash(h);
+
+    mkString(v, hash, context);
+};
+
+
 /*************************************************************
  * Versions
  *************************************************************/
@@ -1135,6 +1159,8 @@ void EvalState::createBaseEnv()
     addPrimOp("__unsafeDiscardStringContext", 1, prim_unsafeDiscardStringContext);
     addPrimOp("__unsafeDiscardOutputDependency", 1, prim_unsafeDiscardOutputDependency);
 
+    addPrimOp("__hash", 2, prim_hash);
+
     // Versions
     addPrimOp("__parseDrvName", 1, prim_parseDrvName);
     addPrimOp("__compareVersions", 2, prim_compareVersions);
